### ISSUE TITLE
[tune] Custom Torch TB Logger

### ISSUE THREAD

**Issue by simofoti** [ray-project/ray#6151]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 18.04
- **Ray installed from (source or binary)**: binary
- **Ray version**: 0.7.6
- **Python version**: 3.6.8
- **Exact command to reproduce**:

### Describe the problem

Hello,

I am not sure this is the place where I am supposed to ask my questions, sorry in case it isn't. The issue I am rising is something in between a feature request (in case you think it's interesting to implement what I am trying to achieve), a possible bug, or me doing something really wrong. 

I'll start by saying that I am trying to write a custom logger to log with **torch.utils.tensorboard**. As you can see below (1), I basically replicated the tf2 logger implementation with some minor adjustment. I am able to successfully log a scalar, but not the *hparams*. Trying to debug I noticed that my *self.trial.evaluated_params* dictionary is always empty. Below you can find also how I set up the PTB scheduler (2) and the parameters used in *tune.run* (3). Do you have any guess why it is empty?

In addition I would love to log some more information like images or videos, but looks like I can't just set a multidimensional tensor as one of the dictionary values returned during the _*train()* step of my *Trainable* class. Do you have any suggestion on how to log this type of information properly? The custom logger sounds like the right place of doing it (ideologically).

Thanks in advance for your help.

### Source code / logs

**(1)** My drafted re-implementation of the logger: 
```python
class CustomLoggerTorch(tune_logger.Logger):
    def __init__(self, config, logdir, trial=None):
        super().__init__(config, logdir, trial)
        self._hp_logged = False
        self._file_writer = None

    def _init(self):
        pass

    def on_result(self, result):
        if self._file_writer is None:
            self._file_writer = SummaryWriter(self.logdir)
        step = result.get(TIMESTEPS_TOTAL) or result[TRAINING_ITERATION]
        tmp = result.copy()
        if not self._hp_logged:
            if self.trial and self.trial.evaluated_params:
                # try:
                self._file_writer.add_hparams(self.trial.evaluated_params,
                                              self.trial.trial_id)
                # except Exception as exc:
                #     tune_logger.error("HParams failed with %s", exc)
            self._hp_logged = True

        for k in ["config", "pid", "timestamp",
                  TIME_TOTAL_S, TRAINING_ITERATION]:
            if k in tmp:
                del tmp[k]  # not useful to log these

        flat_result = flatten_dict(tmp, delimiter="/")
        path = ["ray", "tune"]
        for attr, value in flat_result.items():
            if type(value) in VALID_SUMMARY_TYPES:
                self._file_writer.add_scalar("/".join(path + [attr]),
                                             value, step)
        self._file_writer.flush()

    def flush(self):
        if self._file_writer is not None:
            self._file_writer.flush()

    def close(self):
        if self._file_writer is not None:
            self._file_writer.close()

```
If you think this is of any use, feel free to add it to the repo (once it works).

**(2)** PTB scheduler usage:
```python 
    pbt = PopulationBasedTraining(
        time_attr="training_iteration",
        metric="inception_score",
        mode="max",
        perturbation_interval=3,
        hyperparam_mutations={
            "g_lr": lambda: random.uniform(1e-6, 1e-2),
            "d_lr": lambda: random.uniform(1e-6, 1e-2),
        })
```
I also tried with
```python
hyperparam_mutations={
            "g_lr": [1e-2, 5e-3, 1e-3, 5e-4, 1e-4, 5e-5, 1e-5, 5e-6, 1e-6],
            "d_lr": [1e-2, 5e-3, 1e-3, 5e-4, 1e-4, 5e-5, 1e-5, 5e-6, 1e-6],
        }
```

**(3)** parameters of *tune.run*:
```python
analysis = tune.run(
        TrainGAN,
        name="tune_pbt",
        local_dir=os.path.join(opts.output_path + "/logs", model_name),
        scheduler=pbt,
        loggers=[CustomLoggerTorch],
        reuse_actors=False,
        verbose=True,
        resources_per_trial={
            "cpu": 10,
            "gpu": 1
        },
        stop={
            "training_iteration": 100,
        },
        num_samples=6,
        config={
            "args": opts,
            "params": param,
            "g_lr": 0.0001,
            "d_lr": 0.0001
        })
```

**Comment by richardliaw** [comment]
Ah I was just working on something similar - #6133. The reason why no hparams show up is because we only set `evaluated_params` if it is varied (i.e., via a search algorithm, grid search, or `sample_from`). 

I'm also not quite sure if hparams make sense for PBT because the parameters are updated throughout training, and hparams does not have a time dimension.

**Comment by richardliaw** [comment]
RE: Histogram and images, maybe take a look at this impl for some ideas? 

https://github.com/numenta/nupic.research/blob/master/nupic/research/frameworks/pytorch/tf_tune_utils.py

Let me know if you get something working.

**Comment by simofoti** [comment]
Hi, @richardliaw,

Thanks for your quick response and for your help.

It makes sense that `evaluated_params` is empty with PTB actually. Thanks for pointing that out!

I had a look at your tbx implementation and looks really similar to what I was doing. I think that might make more sense to use **torch.utils.tensorboard** (or at least to implement both) because is embedded in pytorch and with the new release you can also log the hparams (which I don't think is implemented with tbx). What do you think?

It looks like I successfully managed to log videos thank to your precious advice, I need some more testing, but so far so good.

I was also trying to log the current lr for debugging and to be sure it is changed (which as you can see is the only parameter I am mutating), but it is harder than I thought. Accessing it from the optimiser I get always the same value (because I think it is the initial one). 
- Do you have any suggestion on how to do that? 
- For this purpose, do I have to save and restore the optimiser state dictionary?

Thanks again for your help.
Simone

**Comment by richardliaw** [comment]
> I had a look at your tbx implementation and looks really similar to what I was doing. I think that might make more sense to use torch.utils.tensorboard (or at least to implement both) because is embedded in pytorch and with the new release you can also log the hparams (which I don't think is implemented with tbx). What do you think?

I actually got the hparams to work for tbx (see `def close`). In general, it looks like it is the tbx maintainer that is in charge of torch.utils.tensorboard. 

Are you setting the learning rate correctly? You'll have to do something like:
```
    for param_group in optimizer.param_groups:
        param_group['lr'] = lr
```
For the optimizer. Also, you'll want to preserve the optimizer state dict in your trainable. 

I think the `lr` on the logger side is already logged via `ray/tune/config/lr`.

**Comment by simofoti** [comment]
Hi @richardliaw,

I didn't know that the tbx maintainer is also in charge of torch.utils.tensorboard, cool! I'm also glad you managed to log the hparams with tbx :)

I now managed to log the lr (it wasn't automatically logged for me)! I think I was accessing it properly, but I forgot to override `reset_config`.
Now everything seems to be working, but I noticed that I obtain weird log plots because sometimes the step (x-axis) goes "back in time" (see pictures below). In addition the step size is not reported when I save videos, but I suspect this might be a  torch.utils.tensorboard bug.

![logs](https://user-images.githubusercontent.com/44055121/68858463-0171dd80-06dd-11ea-9658-93a09e7edb79.png)
![Screenshot from 2019-11-14 12-44-33](https://user-images.githubusercontent.com/44055121/68863753-dc369c80-06e7-11ea-802a-5959e0bc54fb.png)

Am I doing something wrong? Below you can find my custom logger and the trainable.

```python
class TrainTuningGAN(tune.Trainable):
    def _setup(self, config):
        self.params = config.get('params')
        use_cuda = self.params['use_gpu'] and torch.cuda.is_available()
        self.max_iter = self.params['tune']['max_iter_per_train']
        self.device = torch.device('cuda' if use_cuda else 'cpu')

        self.train_loader_i, self.train_loader_m = \
            utils.get_all_data_loaders(self.params)

        self.trainer = GANTrainer(self.device, self.params).to(self.device)
        self.inception = InceptionScoreNetwork(self.params).to(self.device)

        beta1 = self.params['beta1']
        beta2 = self.params['beta2']
        gen_params = list(self.trainer.gen.parameters())
        dis_params = list(self.trainer.dis.parameters())

        self.g_lr = config.get('g_lr')
        self.d_lr = config.get('d_lr')

        self.gen_opt = \
            torch.optim.Adam([p for p in gen_params if p.requires_grad],
                             lr=self.g_lr,
                             betas=(beta1, beta2),
                             weight_decay=self.params['weight_decay'])
        self.dis_opt = \
            torch.optim.Adam([p for p in dis_params if p.requires_grad],
                             lr=self.d_lr,
                             betas=(beta1, beta2),
                             weight_decay=self.params['weight_decay'])

    def _train(self):
        self.__training_procedure()
        inception_score = self.__testing_procedure()[0]
        batched_videos = self.__visualization_procedure().cpu()
        # print(f"{self.gen_opt.param_groups[0]['lr']} and {self.g_lr}")
        # print(f"{self.dis_opt.param_groups[0]['lr']} and {self.d_lr}")
        return {'inception_score': inception_score,
                'video_out': batched_videos,
                'gen_lr': self.g_lr,
                'disc_lr': self.d_lr}

    def _save(self, tmp_checkpoint_dir):
        checkpoint_path = os.path.join(tmp_checkpoint_dir, 'model.pt')
        torch.save({'gen': self.trainer.gen.state_dict(),
                    'dis': self.trainer.dis.state_dict()}, checkpoint_path)
        checkpoint_opt = os.path.join(tmp_checkpoint_dir, 'optimizer.pt')
        torch.save({'gen': self.gen_opt.state_dict(),
                    'dis': self.dis_opt.state_dict()}, checkpoint_opt)
        return checkpoint_path

    def _restore(self, checkpoint):
        self.trainer.gen.load_state_dict(torch.load(checkpoint)['gen'])
        self.trainer.dis.load_state_dict(torch.load(checkpoint)['dis'])

        checkpoint_opt = checkpoint[:-8] + 'optimizer.pt'
        self.gen_opt.load_state_dict(torch.load(checkpoint_opt)['gen'])
        self.dis_opt.load_state_dict(torch.load(checkpoint_opt)['dis'])

    def reset_config(self, new_config):
        self.g_lr = new_config['g_lr']
        self.d_lr = new_config['d_lr']
        return True

class CustomLoggerTorch(tune_logger.Logger):
    def __init__(self, config, logdir, trial=None):
        super().__init__(config, logdir, trial)
        self._file_writer = None

    def _init(self):
        pass

    def on_result(self, result):
        if self._file_writer is None:
            self._file_writer = SummaryWriter(self.logdir)
        step = result.get(TIMESTEPS_TOTAL) or result[TRAINING_ITERATION]
        tmp = result.copy()

        for k in ["config", "pid", "timestamp",
                  TIME_TOTAL_S, TRAINING_ITERATION]:
            if k in tmp:
                del tmp[k]  # not useful to log these

        if 'video_out' in tmp:
            self._file_writer.add_video('video', tmp['video_out'],
                                        walltime=step)
            del tmp['video_out']

        flat_result = flatten_dict(tmp, delimiter="/")
        path = ["ray", "tune"]
        for attr, value in flat_result.items():
            if type(value) in VALID_SUMMARY_TYPES:
                self._file_writer.add_scalar("/".join(path + [attr]),
                                             value, step)
        self._file_writer.flush()

    def flush(self):
        if self._file_writer is not None:
            self._file_writer.flush()

    def close(self):
        if self._file_writer is not None:
            self._file_writer.close()


```

Thanks again for your precious help!
Simone

**Comment by simofoti** [comment]
I've just found that this "problem" was already reported in #5312 and according to #5445 looks like it hasn't been merged yet. Am I correct? Are you planning to (nightly) release it soon?

According to what I found, I assume that there is nothing wrong with my code. Right? :)

Thanks again,
Simone

**Comment by richardliaw** [comment]
Ah, there's nothing wrong with your code. I need to restructure PBT but haven't gotten it yet. I can probably get to this in a week or so.

**Comment by simofoti** [comment]
Oh nice, I'm relieved to hear that my code is correct! I hope to get some decent result even without a proper logging. I'll wait for the updated version of PBT.

May I also ask your personal advice on the best scheduler implemented in tune? Do you think PBT is a good choice or do you recommend something else?

**Comment by richardliaw** [comment]
PBT makes a lot of sense if you want to optimize over a "schedule" of 1 or 2 parameters, like learning rate or something. 

Otherwise, I generally like just using ASHA.

**Comment by simofoti** [comment]
@richardliaw thanks for your help and all your precious suggestions! I'll try also ASHA when I'll need to tune some more parameter :)
I will keep track of the PBT restructuring from the other issue, so I think that we can close this one.
